[ttk] Add access to logged events in BehaviorTestKit #25905

jlprat opened the pull request:
`BehaviorTestKit` now has methods to:
 - Access to logged events
 - Clear all log events

Added section in docs explaining how to check for logs.

Note about the implementation: `CapturedLogEvent` was effectively private to akka. It uses `OptionVal` internally, this makes it impossible to be used from outside akka. To overcome this, I added some methods to translate it to the corresponding `Option` and `Optional`.

jlprat committed:
[ttk] Add access to logged events in BehaviorTestKit #25905

`BehaviorTestKit` now has methods to:
 - Access to logged events
 - Clear all log events

Added section in docs explaining how to check for logs.

Note about the implementation: `CapturedLogEvent` was effectively private to akka. It uses `OptionVal` internally, this makes it impossible to be used from outside akka. To overcome this, I added some methods to translate it to the corresponding `Option` and `Optional`.

jlprat commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Representation of a Log Event issued by a [[akka.actor.typed.Behavior]]
+ */
+final case class CapturedLogEvent private[akka] (logLevel: LogLevel, message: String,
```
I guess this class could be refactored to the Java API one, and then the Scala one that implements the first one.

jlprat commented:
refs: #25905

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala:
```diff
@@ -0,0 +1,66 @@
+                                                 mdc:                Map[String, Any]) {
+
+  /**
+   * Constructor for Java code
```
the convention is "Java API"

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala:
```diff
@@ -0,0 +1,66 @@
+import akka.actor.typed.LogMarker
+import akka.event.Logging.LogLevel
+import akka.util.OptionVal
+import akka.util.OptionVal.{ None, Some }
```
I'd rather use these as `OptionVal.Some` and `OptionVal.None` to avoid confusing them with standard ones

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala:
```diff
@@ -0,0 +1,66 @@
+
+import scala.collection.JavaConverters._
+import scala.compat.java8.OptionConverters._
+import scala.{ None ⇒ SNone, Option ⇒ SOption, Some ⇒ SSome }
```
and then this can be removed

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala:
```diff
@@ -0,0 +1,66 @@
+ */
+final case class CapturedLogEvent private[akka] (logLevel: LogLevel, message: String,
+                                                 private val cause:  OptionVal[Throwable],
+                                                 private val marker: OptionVal[LogMarker],
```
I don't think we need to use OptionVal in testkit, it's intended as a performance optimization

jlprat replied:
OK, then I'll use Scala's `Option` in here and do the `OptionVal` to `Option` conversion in here: https://github.com/akka/akka/blob/master/akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/StubbedActorContext.scala#L76

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/BehaviorTestKitImpl.scala:
```diff
@@ -138,4 +137,10 @@ private[akka] final class BehaviorTestKitImpl[T](_path: ActorPath, _initialBehav
+
+  override def getAllLogEntries(): util.List[CapturedLogEvent] = logEntries().asJava
+
+  override def logEntries(): Seq[CapturedLogEvent] = context.logEntries
```
`immutable.Seq`

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/BehaviorTestKit.scala:
```diff
@@ -129,7 +129,17 @@ trait BehaviorTestKit[T] {
+  /**
+   * Returns all the [[CapturedLogEvent]] issued by this behavior(s)
+   */
+  def logEntries(): Seq[CapturedLogEvent]
```
`immutable.Seq`

patriknw reviewed (commented):
approach is looking good, some nitpicks

jlprat committed:
Apply feedback

Clean up `CapturedLogEvent` by using only `Option` in public methods
Use `immutable.Seq` instead of plain `Seq`

jlprat commented:
Feedback is applied. Ready to review whenever you have time.

chbatey commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Representation of a Log Event issued by a [[akka.actor.typed.Behavior]]
+ */
+final case class CapturedLogEvent(
```
Optional. Add some extra constructors to not pass in errorCause, marker and mdc for the common case of these not being set.

jlprat replied:
I will do later today

jlprat replied:
As it got merged already I will create a new PR with this change :slightly_smiling_face:

patriknw replied:
thanks

jlprat replied:
Created https://github.com/akka/akka/pull/26016

chbatey reviewed (approved):
LGTM, thanks

patriknw reviewed (approved):
LGTM, thanks!
